services/galley/src/Galley/API/Message.hs

```diff
@@ -213,26 +220,110 @@ postQualifiedOtrMessage senderType sender mconn convId msg = runUnionT $ do
   unless sendMessage $ do
     lift $ guardQualifiedLegalholdPolicyConflicts (userToProtectee senderType sender) (qmMissing mismatch)
     throwUnion $ WithStatus @412 otrResult
-  let qualifiedConv = Qualified convId localDomain
-      qualifiedSender = Qualified sender senderDomain
-      localValidMessages = Map.mapKeys (\(_, u, c) -> (u, c)) $ Map.filterWithKey (\(d, _, _) _ -> localDomain == d) validMessages
-      metadata = qualifiedNewOtrMetadata msg
-      events =
-        localValidMessages & reindexed snd itraversed
+  failedToSend <-
+    lift $
+      sendMessages
+        now
+        (Qualified sender senderDomain)
+        senderClient
+        mconn
+        convId
+        localMemberMap
+        (qualifiedNewOtrMetadata msg)
+        validMessages
+  throwUnion $ WithStatus @201 (otrResult {mssFailedToSend = mkQualifiedUserClients failedToSend})
+  where
+    makeUserMap :: Set UserId -> Map UserId (Set ClientId) -> Map UserId (Set ClientId)
+    makeUserMap keys = (<> Map.fromSet (const mempty) keys)
+
+-- | Send both local and remote messages, return the set of clients for which
+-- sending has failed.
+sendMessages ::
+  UTCTime ->
+  Qualified UserId ->
+  ClientId ->
+  Maybe ConnId ->
+  ConvId ->
+  Map UserId LocalMember ->
+  MessageMetadata ->
+  Map (Domain, UserId, ClientId) ByteString ->
+  Galley (Set (Domain, UserId, ClientId))
+sendMessages now sender senderClient mconn conv localMemberMap metadata messages = do
+  localDomain <- viewFederationDomain
+  let messageMap = Map.foldrWithKey (\(d, u, c) t -> Map.insertWith (<>) d (Map.singleton (u, c) t)) mempty messages
+  fmap (setOf ((itraversed <. folded) . withIndex . to (\(d, (u, c)) -> (d, u, c)))) . flip Map.traverseWithKey messageMap $ \d msgs ->
+    if d == localDomain
+      then do
+        sendLocalMessages now sender senderClient mconn conv localMemberMap metadata msgs
+        pure mempty
+      else do
+        x <- sendRemoteMessages d now sender senderClient conv metadata msgs
+        case x of
+          Nothing -> pure mempty
+          Just e -> do
```

Comment:
This code could be refactored a little for improved readability using smaller named helper functions. the lines 253 and 254 look a little bit like they partition the map of domain/user/clients into local and remote, so perhaps a helper function could be created here?

Reply:
I moved the local/remote distinction logic to a separate function with a guard, and inline the remote error handling in the `sendMessagesRemote` function.

I've also used the opportunity to get rid of the extra flattening/unflattening of `failedToSend`.

Hopefully things are a bit clearer now.

Reply:
Thank you! Looks more readable now.